plugins/filesystem/local/adapter/adapter.php

```
@@ -326,4 +326,272 @@ private function getDate($date = null)
+	 * @param   bool    $force            Set true to overwrite file if exists
+	 *
+	 * @return bool
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	protected function moveFile($sourcePath, $destinationPath, $force = false)
+	{
+		if (!file_exists($destinationPath))
+		{
+			$success  = JFile::move($sourcePath, $destinationPath);
+			return $success;
+		}
+		else
+		{
+			if ($force && !is_dir($destinationPath))
+			{
+				$success  = JFile::move($sourcePath, $destinationPath);
+				return $success;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Moves a folder from source to destination
+	 *
+	 * @param   string  $sourcePath       Source path of the file or directory
+	 * @param   string  $destinationPath  Destination path of the file or directory
+	 * @param   bool    $force            Set true to overwrite files or directories
+	 *
+	 * @return bool
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	protected function moveFolder($sourcePath, $destinationPath, $force = false)
+	{
+		if (file_exists($destinationPath))
+		{
+			if (is_dir($destinationPath))
+			{
+				// We need to bypass exception thrown in JFolder when destination exists
+				// So we only copy it in forced condition, then delete the source to simulate a move
+				if ($force)
+				{
+					$copy_success = JFolder::copy($sourcePath, $destinationPath, '', $force);
+					$delete_success = false;
+
+					if ($copy_success)
+					{
+						$delete_success = JFolder::delete($sourcePath);
+					}
+
+					return $copy_success && $delete_success;
+				}
+			}
+			else
+			{
+				// Sometimes a file with destination path could exists
+				// If forced we can delete it and move folder
+				if ($force)
+				{
+					$delete_success = JFile::delete($destinationPath);
+					$move_success = false;
+					if ($delete_success)
+					{
+						$move_success = JFolder::move($sourcePath, $destinationPath);
+					}
+
+					return self::convertToBool($move_success) && $delete_success;
```

IS this here really needed convertToBool?

---

Yes, as in move it returns a error text when failing

---

I used this function because of this in JFolder move method

         

> /**
> 	 * Moves a folder.
> 	 *
> 	 * @param   string   $src          The path to the source folder.
> 	 * @param   string   $dest         The path to the destination folder.
> 	 * @param   string   $path         An optional base path to prefix to the file names.
> 	 * @param   boolean  $use_streams  Optionally use streams.
> 	 *
> 	 * @return  mixed  Error message on false or boolean true on success.
> 	 *
> 	 * @since   11.1
> 	 */

https://github.com/joomla-projects/media-manager-improvement/blob/milestone1/libraries/joomla/filesystem/folder.php

---

Them we do it like in the other functions like delete. Return type is void and when something goes wrong an exception will be thrown.

---

I looked at how delete implemented in the adapter, actually the JFolder::delete returns a boolean, true or false on success, in our case move returns a true or mixed error message, in that case we want to check whether it was a boolean to perform, error message will be also taken as true in php

---

If an error message is returned, you need to throw an exception.